Add file to generate peer factor

KatalinaBobowik opened the pull request:
This PR is moved from the the [ancestry repository](https://github.com/populationgenomics/ancestry) (see the history [here](https://github.com/populationgenomics/ancestry/commits/peer/scripts/hail_batch/peer)).

KatalinaBobowik commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,139 @@
+    return covariates.to_numpy()
+
+
+def run_peer(expression_file, covariates, factors_output_path):
```
This is currently failing because dill is not installed. See the failed log [here](https://batch.hail.populationgenomics.org.au/batches/8273/jobs/2).

illusional replied:
Hoping https://github.com/populationgenomics/images/pull/4/files will unblock this PR.

illusional replied:
Hoping https://github.com/populationgenomics/images/pull/4 will unblock this PR.

illusional commented:
Just noting this is blocked because Hail Batch is trying to run the python job in Python 3, but peer is not installed for python 3:

- Can we make hail batch run in python 2 for this job?
- Can we install Peer for python 3?

KatalinaBobowik commented:
@illusional I don't think we can install Peer for python 3. I tried to look into running hail batch in pyhon 2 by installing a python 2.7 conda environment, then trying to run hail batch. Unfortunately, this doesn't seem to work. Alternatively, I could run the script in R. Would this solve any problems?

Commit:
Moving from ancestry branch to tob-wgs branch

See past commits here: https://github.com/populationgenomics/ancestry/tree/peer/scripts/hail_batch/peer

Commit:
Fix boolean logic for pylint error

Fixing error 'E712 comparison to False should be 'if cond is False:' or 'if not cond:''

Commit:
Fxing error 'at least two spaces before inline comment'

Commit:
Remove cpu request, as this does nothing to the run time

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+"""
+Test run this with:
+
+    cd <dir where peer.py file is>
```
Can probably update this now that the path has been decided:

```suggestion
    cd scripts/peer
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+from google.cloud import storage
+
+PEER_DOCKER = 'australia-southeast1-docker.pkg.dev/cpg-common/images/peer:1.3.2'
+SCORES_PATH = 'gs://cpg-tob-wgs-test/kat/pca/nfe_feb22/v0/scores.json'
+COVARIATES_PATH = 'gs://cpg-tob-wgs-test/scrna-seq/grch38_association_files/covariates_files/covariates.tsv'
+SAMPLE_ID_KEYS_PATH = 'gs://cpg-tob-wgs-test/scrna-seq/grch38_association_files/metadata/keys_metadata_sheet.csv'
```
These constants are good for testing, but it would probably be better to remove them, and always specify the files when you run this script.

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    scores.insert(loc=0, column='sampleid', value=sampleid)
+    # filter to only CPG samples
+    scores = scores[scores.sampleid.str.contains('CPG')]
+    print(len(scores.index))
```
Left-over debug print, otherwise these should get turned into proper logging with a message (and using the logging library).

KatalinaBobowik replied:
deleted the print message

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    scores = pd.merge(
+        scores, sampleid_keys, how='left', left_on='sampleid', right_on='CPG_ID'
+    )
+    print(len(scores.OneK1K_ID.dropna()))
```
ditto, I'll stop highlighting them but there are a couple more.

KatalinaBobowik replied:
I've used the logging library as well

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+
+def get_covariates(
+    scores_path, covariates_path, expression_file, sample_id_keys_path
+) -> str:
```
This doesn't match the return type

KatalinaBobowik replied:
I'm a little confused on what would go here instead, since we have two dataframes being returned

KatalinaBobowik replied:
as you mentioned below, would it look like this?
`-> Tuple[str, str]`

KatalinaBobowik replied:
I'm getting a pylint error:
`undefined name 'Tuple'`. Not sure if this should be lowercase, or something else?

illusional replied:
You'll need to import it from typing:

```python
from typing import Tuple
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    covariates[['sex', 'age']] = covariates[['sex', 'age']].astype(int)
+
+    # return expression data and covariates
+    return covariates.to_csv(index=False), expression.to_csv(index=False)  # type: ignore
```
Can we remove the type ignore statement - the initial type annotation in the function is incorrect (it should be Tuple[str, str]). Otherwise can we add a comment on why it's ignored?

KatalinaBobowik replied:
I've removed the type ignore statement after updating the return type

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+
+def get_at_index(obj, idx):
+    """
+    get object index
```
```suggestion
    Get the object at index, required to break up the result
    from the get_covariates data within Hail Batch    
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+
+    job.image(PEER_DOCKER)
+
+    job.command(f'echo "expressions file:" && head {expression_file}')
+    job.command(f'echo "covariates file:" && head {covariates_file}')
```
Can remove these extra print debug statements now

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    Get covariate data for each cell type
+    \"""
+
+    print 'Loading data'
```
Should remove these debug statements, or make them more descriptive

KatalinaBobowik replied:
I removed them

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+
+if __name__ == "__main__":
+    import sys
+    run_peer(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5], sys.argv[6])
```
Maybe could simplify this with:

```suggestion
    run_peer(*sys.argv[1:])
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    job.command(
+        f'python run_peer.py {expression_file} {covariates_file} {job.factors_output_path} {job.weights_output_path} {job.precision_output_path} {job.residuals_output_path}'
+    )
+    job.command('ls -l')
```
Debug statement to remove

KatalinaBobowik replied:
removed

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    )
+
+    job.command(
+        f'python run_peer.py {expression_file} {covariates_file} {job.factors_output_path} {job.weights_output_path} {job.precision_output_path} {job.residuals_output_path}'
```
It's probably worth breaking these up on different lines, you won't run into line length issues, and it'll be easier to debug the command in Hail Batch:

```suggestion
        f'python run_peer.py \
            {expression_file} \
            {covariates_file} \
            {job.factors_output_path} \
            {job.weights_output_path} \
            {job.precision_output_path} \
            {job.residuals_output_path}'
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    batch: hb.Batch,
+    cell_type_name: str,
+    expression_file,
+    scores_path=SCORES_PATH,
+    covariates_path=COVARIATES_PATH,
+    sample_id_keys_path=SAMPLE_ID_KEYS_PATH,
```
As mentioned before, I think these should not be hard-coded anymore, but always supplied.

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    assert path_to_cell_files.startswith('gs://')
+    cs_client = storage.Client()
+
+    bucket_name = path_to_cell_files.split('gs://')[1].split('/')[0]
```
```suggestion
    bucket_name = path_to_cell_files.lstrip('gs://').split('/')[0]
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+
+    bucket_name = path_to_cell_files.split('gs://')[1].split('/')[0]
+    bucket = cs_client.get_bucket(bucket_name)
+    bucket_path = path_to_cell_files.split(f'gs://{bucket_name}/')[-1]
```
Another couple of ways that don't rely on split:

```suggestion
    bucket_path = path_to_cell_files.lstrip('gs://{bucket_name}/')
    bucket_path = path_to_cell_files[(5 + len(bucket_name) + 1):]
```

KatalinaBobowik replied:
Maybe I missed it, but why should we be avoiding using `split`?

KatalinaBobowik replied:
I used the second example but am getting the error:
`Using .strip() with multi-character strings is misleading the reader. It looks like stripping a substring. Move your character set to a constant if this is deliberate. Use .replace() or regular expressions to remove string fragments.`

KatalinaBobowik replied:
(from pylint)

illusional replied:
Ah okay, let's not do that :) `.strip` actually doesn't work in the way I expected it to. Could you revert this back to your original code?

KatalinaBobowik replied:
np, will do that now

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    bucket_path = path_to_cell_files.split(f'gs://{bucket_name}/')[-1]
+
+    logging.info(f'Going to fetch cell types from {bucket_path}')
+    blobs = bucket.list_blobs(prefix=bucket_path)
```
This is important as otherwise it gets all files in all subdirectories of the path.

```suggestion
    blobs = bucket.list_blobs(prefix=bucket_path, delimiter='/')
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    return job
+
+
+# @click.command()
+# @click.option('--expression-file')
```
```suggestion
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    """
+
+    dataset = os.getenv('CPG_DATASET')
+    driver_image = os.getenv('CPG_DRIVER_IMAGE')
```
You'll want an assert here to make sure the driver_image is set:
```suggestion
    driver_image = os.getenv('CPG_DRIVER_IMAGE')
    assert driver_image
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+    cell_types: list = find_cell_types_from_path(path_to_cell_files)
+
+    for cell_type in cell_types:
+        expression_file = f'gs://cpg-tob-wgs-test/scrna-seq/grch38_association_files/expression_files/{cell_type}_expression.tsv'
```
Can you use the `path_to_cell_files` to construct this path - otherwise this should be passed in as an input to the script (this is linking to the test bucket atm)

KatalinaBobowik replied:
good catch, added `path_to_cell_files` instead

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+        expression_file = f'gs://cpg-tob-wgs-test/scrna-seq/grch38_association_files/expression_files/{cell_type}_expression.tsv'
+        process_cell_type_on_batch(batch, cell_type, expression_file=expression_file)
+
+        # batch.run(dry_run=True)
```
```suggestion
```

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,273 @@
+
+@click.command()
+@click.option('--path-to-cell-files')
+def main(path_to_cell_files):
```
You'll now want to add an option for each scores_path, covariates_path, sample_id_keys_path, and connect them when you call `process_cell_type_on_batch` below.

illusional reviewed (changes requested):
There are a few debug statements left in the script, and a couple of variables that IMO should be parameterised, but looks really great!

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Remove print statements

Commit:
Merge branch 'peer' of github.com:populationgenomics/tob-wgs into peer

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Add in recommendations made by Michael

See https://github.com/populationgenomics/tob-wgs/pull/80

Commit:
Parameterise scores, covariates, and sample ID keys

Commit:
Update docstring to reflect newly-parameterised files

Commit:
Remove outdated, hard-coded variables (now parameterised)

Commit:
Add in Tuple module for defining return type

illusional commented on scripts/peer/peer.py:
```diff
@@ -0,0 +1,272 @@
+    )
+
+    job.command(
+        f'python run_peer.py \
+            {expression_file} \
+            {covariates_file} \
+            {job.factors_output_path} \
+            {job.weights_output_path} \
+            {job.precision_output_path} \
+            {job.residuals_output_path}'
```
My suggestion here wasn't correct, this should be a multiline string (rather than just adding new lines), sorry this suggestion should be better:

```suggestion
        f"""python run_peer.py \
            {expression_file} \
            {covariates_file} \
            {job.factors_output_path} \
            {job.weights_output_path} \
            {job.precision_output_path} \
            {job.residuals_output_path}"""
```

Commit:
Fix pylint issue with lstrip

Fixing error '.strip() with multi-character strings is misleading the reader'

Commit:
Update scripts/peer/peer.py

Co-authored-by: Michael Franklin <michael@illusional.net>

Commit:
Fix pylint issue 'undefined name 'bucket_name''

I replaced the wrong line when fixing the 'strip() with multi-character strings is misleading the reader.' pylint issue. Reverting back to original code
